structuredevent-service-cdp/src/main/java/com/sequenceiq/cloudbreak/structuredevent/rest/CDPRestCommonService.java

```diff
@@ -0,0 +1,129 @@
+import com.sequenceiq.cloudbreak.structuredevent.event.rest.RestRequestDetails;
+
+@Component
+public class CDPRestCommonService {
+
+    public static final String NAME_PATH = "name";
+
+    public static final String NAMES_PATH = "names";
+
+    public static final String RESOURCE_CRN_PATH = "crn";
+
+    public static final String CRNS_PATH = "crns";
+
+    public Map<String, String> addClusterCrnAndNameIfPresent(RestCallDetails restCallDetails, Map<String, String> restParams, String nameField,
+            String crnField) {
+        Map<String, String> params = new HashMap<>();
+        RestRequestDetails restRequest = restCallDetails.getRestRequest();
+        Json requestJson = getJson(restRequest.getBody());
+        Json responseJson = getJson(restCallDetails.getRestResponse().getBody());
+        String resourceCrn = getCrn(requestJson, responseJson, restRequest, restParams, crnField);
+        String name = getName(requestJson, responseJson, restRequest, restParams, nameField);
+
+        checkNameOrCrnProvided(restRequest, resourceCrn, name);
+
+        if (StringUtils.isNotEmpty(name)) {
+            params.put(nameField, name);
+        }
+
+        if (StringUtils.isNotEmpty(resourceCrn)) {
+            params.put(crnField, resourceCrn);
+        }
+        return params;
+    }
+
+    private String getName(Json requestJson, Json responseJson, RestRequestDetails request, Map<String, String> restParams, String nameField) {
+        if (StringUtils.isEmpty(restParams.get(nameField))) {
+            return getResourceId(requestJson, responseJson, request, NAME_PATH, NAMES_PATH, restParams, "name");
+        }
+        return restParams.get(nameField);
+    }
+
+    private String getCrn(Json requestJson, Json responseJson, RestRequestDetails request, Map<String, String> restParams, String crnField) {
+        if (StringUtils.isEmpty(restParams.get(crnField))) {
+            return getResourceId(requestJson, responseJson, request, RESOURCE_CRN_PATH, CRNS_PATH, restParams, "crn");
+        }
+        return restParams.get(crnField);
+    }
+
+    private String getResourceId(Json requestJson, Json responseJson, RestRequestDetails request, String path, String pluralPath, Map<String, String> restParams,
+            String idType) {
+        String id = null;
+        if (requestJson != null) {
+            id = getValueFromJson(requestJson, path);
+            if (StringUtils.isEmpty(id)) {
+                id = getListValue(request, requestJson, pluralPath, restParams, idType);
+            }
+        }
+        if (responseJson != null && StringUtils.isEmpty(id)) {
+                id = Optional.ofNullable(getValueFromJson(responseJson, path))
+                        .orElse(getListValue(request, responseJson, pluralPath, restParams, idType));
+        }
+        return id;
+    }
+
+    private String getListValue(RestRequestDetails restRequest, Json json, String path, Map<String, String> restParams, String idType) {
```

Comment:
it's a bit hard to follow this method, similarly to `getResourceId`. Could you refactor this a bit to make it more readable?